hibernate-core/src/main/java/org/hibernate/sql/ast/tree/insert/InsertStatement.java

```diff
@@ -7,60 +7,73 @@
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 
 import org.hibernate.sql.ast.tree.MutationStatement;
+import org.hibernate.sql.ast.tree.cte.CteConsumer;
 import org.hibernate.sql.ast.tree.expression.ColumnReference;
-import org.hibernate.sql.ast.tree.expression.Expression;
 import org.hibernate.sql.ast.tree.from.TableReference;
+import org.hibernate.sql.ast.tree.select.QuerySpec;
+
+import org.jboss.logging.Logger;
 
 /**
  * @author Steve Ebersole
  */
-public class InsertStatement implements MutationStatement {
+public class InsertStatement implements MutationStatement, CteConsumer {
+	private static final Logger log = Logger.getLogger( InsertStatement.class );
+
 	private TableReference targetTable;
 	private List<ColumnReference> targetColumnReferences;
-	private List<Expression> values;
-
-	public InsertStatement(TableReference targetTable) {
-		this.targetTable = targetTable;
-	}
+	private QuerySpec sourceSelectStatement = null;
+	private List<Values> valuesList = new ArrayList<>();
 
 	public TableReference getTargetTable() {
 		return targetTable;
 	}
 
 	public void setTargetTable(TableReference targetTable) {
+		log.tracef( "Setting INSERT target table [%s]", targetTable );
+		if ( this.targetTable != null ) {
+			log.debugf( "INSERT target table has been set multiple times" );
+		}
 		this.targetTable = targetTable;
 	}
 
 	public List<ColumnReference> getTargetColumnReferences() {
-		return targetColumnReferences;
+		return targetColumnReferences == null ? Collections.emptyList() : targetColumnReferences;
 	}
 
-	public void setTargetColumnReferences(List<ColumnReference> targetColumnReferences) {
-		this.targetColumnReferences = targetColumnReferences;
+	public void addTargetColumnReferences(ColumnReference... references) {
+		if ( targetColumnReferences == null ) {
+			targetColumnReferences = new ArrayList<>();
+		}
+
+		Collections.addAll( this.targetColumnReferences, references );
 	}
 
-	public void addTargetColumnReference(ColumnReference columnReference) {
+	public void addTargetColumnReferences(List<ColumnReference> references) {
 		if ( targetColumnReferences == null ) {
 			targetColumnReferences = new ArrayList<>();
 		}
-		targetColumnReferences.add( columnReference );
+
+		this.targetColumnReferences.addAll( references );
 	}
 
-	public List<Expression> getValues() {
-		return values;
+	public QuerySpec getSourceSelectStatement() {
+		return sourceSelectStatement;
 	}
 
-	public void setValues(List<Expression> values) {
-		this.values = values;
+	public void setSourceSelectStatement(QuerySpec sourceSelectStatement) {
+		this.sourceSelectStatement = sourceSelectStatement;
 	}
 
-	public void addValue(Expression expression) {
-		if ( values == null ) {
-			values = new ArrayList<>();
-		}
-		values.add( expression );
+	public List<Values> getValuesList() {
+		return valuesList;
+	}
+
+	public void setValuesList(List<Values> valuesList) {
+		this.valuesList = valuesList;
```

Comment:
Saw the g4 rule dictates the list shoud not be empty. Do we need to guard against empty list to avoid producing invalid SQL?